### PR TITLE
Feature/hdf5

### PR DESCRIPTION
**Pull request by hroest**
add HDF5 support

we should discuss whether we should make this optional or required. This should work once we have https://github.com/OpenMS/contrib/pull/84 merged.

### CODE REVIEW

**Comment by hroest** [comment]
looks like travis is running through but none of the other ones which would need an update in the contrib first.

**Comment by hroest** [comment]
hmm pyOpenMS has a strange definition of success:

```

CMake Error at /usr/share/cmake-3.2/Modules/FindPackageHandleStandardArgs.cmake:138 (message):
  Could NOT find HDF5 (missing: HDF5_LIBRARIES HDF5_INCLUDE_DIRS)
Call Stack (most recent call first):
  /usr/share/cmake-3.2/Modules/FindPackageHandleStandardArgs.cmake:374 (_FPHSA_FAILURE_MESSAGE)
  /usr/share/cmake-3.2/Modules/FindHDF5.cmake:362 (find_package_handle_standard_args)
  src/openms/cmake_findExternalLibs.cmake:138 (find_package)
  src/openms/CMakeLists.txt:70 (include)


-- Configuring incomplete, errors occurred!
See also "/home/jenkins/ws/openms_pr_docker_pyopenms/OpenMS-build/CMakeFiles/CMakeOutput.log".
See also "/home/jenkins/ws/openms_pr_docker_pyopenms/OpenMS-build/CMakeFiles/CMakeError.log".
make: *** No rule to make target `pyopenms'.  Stop.
   Site: ae9613dd46ec
   Build name: Linux-c++
Create new tag: 20190418-2214 - Experimental
Test project /home/jenkins/ws/openms_pr_docker_pyopenms/OpenMS-build
No tests were found!!!
Archiving artifacts
INFO: Starting to record.
INFO: Processing CTest-Version 3.x (default)
INFO: [CTest-Version 3.x (default)] - 1 test report file(s) were found with the pattern 'OpenMS-build/Testing/**/*.xml' relative to '/home/jenkins/ws/openms_pr_docker_pyopenms' for the testing framework 'CTest-Version 3.x (default)'.
WARNING: All test reports are empty.
INFO: Check 'Failed Tests' threshold.
INFO: Check 'Skipped Tests' threshold.
INFO: Setting the build status to SUCCESS
INFO: Stopping recording.
Finished: SUCCESS
```

**Comment by jpfeuffer** [comment]
So:
1) I fixed the result recording of the job
2) The pyopenms job uses your docker image, so you need to update it, too.
3) On linux it seems like HDF5 depends on libdl which is somehow not found on the machine (A first guess)
4) On windows the Cmake find module does not work correctly, potentially because only the "lib" files exist. (Maybe build with HDF5_USE_STATIC_LIBRARIES?)

**Comment by hroest** [comment]
> On linux it seems like HDF5 depends on libdl which is somehow not found on the machine (A first guess)

strange, on my machine it does the following which means that `libdl` is linked at least twice in there...

../../lib/libOpenSwathAlgo.so -lCbc -lCgl -lClp -lCoinUtils -lOsi -lOsiClp -lsvm -lxerces-c -lboost_iostreams -lboost_date_time -lboost_math_c99 -lboost_regex /home/hr/openmsall/builds/contrib-exp/lib/libWm5Core.a /home/hr/openmsall/builds/contrib-exp/lib/libWm5Mathematics.a -lbz2 -lz -lsqlite3 /home/hr/openmsall/builds/contrib-exp/lib/libglpk.a /usr/lib/x86_64-linux-gnu/libQt5Network.so.5.5.1 /usr/lib/x86_64-linux-gnu/hdf5/serial/lib/libhdf5_cpp.so /usr/lib/x86_64-linux-gnu/hdf5/serial/lib/libhdf5.so -lpthread -lsz -lz **-ldl** -lm -lsqlite3 /home/hr/openmsall/builds/contrib-exp/lib/libglpk.a /usr/lib/x86_64-linux-gnu/hdf5/serial/lib/libhdf5_cpp.so /usr/lib/x86_64-linux-gnu/hdf5/serial/lib/libhdf5.so -lpthread -lsz **-ldl** -lm /usr/lib/x86_64-linux-gnu/libQt5Core.so.5.5.1 -fopenmp -lpthread -Wl,-rpath,/home/hr/openmsall/builds/openms_hdf5/lib:/usr/lib/x86_64-linux-gnu/hdf5/serial/lib: 


> On windows the Cmake find module does not work correctly, potentially because only the "lib" files exist. (Maybe build with HDF5_USE_STATIC_LIBRARIES?)

on Linux, `-DBUILD_SHARED_LIBRARIES=On` builds the shraed .so libraries

> The pyopenms job uses your docker image, so you need to update it, too. (https://github.com/hroest/docker_files ; dependencies-qt5)

would be easy to fix, do you remember which docker container you currently use? `apt-get install -y --no-install-recommends --no-install-suggests libhdf5-dev` should do it

File: src/openms/CMakeLists.txt [hunk 1]
```diff
@@ -93,7 +93,9 @@ set(OPENMS_DEP_LIBRARIES  ${COIN_LIBRARIES}
                           ${BZIP2_LIBRARIES}
                           ${ZLIB_LIBRARIES}
                           ${SQLite_LIBRARY}
+                          ${HDF5_CXX_LIBRARIES}
                           ${GLPK_LIBRARIES}
+                          ${CMAKE_DL_LIBS}
```
Comment (hroest) [inline src/openms/CMakeLists.txt]: strange, but this fixed the openms_pr_ci/compiler=g++-7,os_label=ubuntu_1804 builds...

**Comment by hroest** [comment]
On Windows, I now see the dlls being built:

```
cmake -DBUILD_SHARED_LIBRARIES=On -DBUILD_TYPE=HDF5 -G "Visual Studio 14 2015 Win64" ..\..\source\contrib_test


2019-04-19  09:32 AM    <DIR>          .
2019-04-19  09:32 AM    <DIR>          ..
2019-04-19  09:32 AM         3,009,024 hdf5.dll
2019-04-19  09:32 AM           618,486 hdf5.lib
2019-04-19  09:32 AM           377,344 hdf5_cpp.dll
2019-04-19  09:32 AM           397,230 hdf5_cpp.lib
2019-04-19  09:32 AM           112,128 hdf5_hl.dll
2019-04-19  09:32 AM            27,664 hdf5_hl.lib
2019-04-19  09:32 AM            15,360 hdf5_hl_cpp.dll
2019-04-19  09:32 AM            11,804 hdf5_hl_cpp.lib
2019-04-19  09:32 AM        10,393,572 libhdf5.lib
2019-04-19  09:29 AM             2,943 libhdf5.settings
2019-04-19  09:32 AM         2,356,894 libhdf5_cpp.lib
2019-04-19  09:32 AM           306,412 libhdf5_hl.lib
2019-04-19  09:32 AM            15,912 libhdf5_hl_cpp.lib
2019-04-19  09:32 AM    <DIR>          pkgconfig
```

but only if `-DBUILD_SHARED_LIBRARIES=On ` - should we make this the default?

**Comment by hroest** [comment]
> On Windows, I now see the dlls being built:

it seems that most other libs that we build are static, so we probably also want HDF5 to be statically linked, right? I got this to work with cc69660c8d2e379135df078553de3a7f84dc611a which specifically looks for the static libs and uses those. However, on OSX and Linux we probably want dynamic linkage, right?

**Comment by jpfeuffer** [comment]
Yes right. On Windows most libs are static. If you would do dynamic, we would need to add a copy step for packaging in the KNIME (and maybe NSIS) cmake scripts. On Mac (dynamic) dependencies are calculated and shipped automatically by our script, on Linux the debian package tool will calculate and add the dynamic dependencies as apt packages. Looks good for now. I'll check packaging once it is in the nightlies.

I can add the apt-get manually to the docker container but maybe it is better, if you update the docker scripts so everyone benefits. We have a CI script that is going to pull the Dockerfile and rebuild the container used for PRs. https://github.com/hroest/docker_files/blob/master/openms/dependencies-qt5/Dockerfile

**Comment by hroest** [comment]
>  https://github.com/hroest/docker_files/blob/master/openms/dependencies-qt5/Dockerfile

ok, I updated it and it should work now (not tested)

**Comment by hroest** [comment]
@jpfeuffer can you test the docker container? it should now have the HDF5 libraries.

**Comment by jpfeuffer** [comment]
I restarted the job with the new container. Ignore the non-docker one (started by accident).

**Comment by jpfeuffer** [comment]
Worked. Ready to merge?

**Comment by hendrikweisser** [comment]
The new required package(s) need(s) to be added to the build instructions (e.g. `doc/doxygen/install/install-linux.doxygen`).

**Comment by hroest** [comment]
> The new required package(s) need(s) to be added to the build instructions (e.g. `doc/doxygen/install/install-linux.doxygen`).

true, see #4086

**Pull request merged**